### PR TITLE
Refine the Array.prototype.reduce() doc for clarity

### LINKED ISSUE #6156: Issue with "Array.prototype.reduce()": (conflicting info on usage)
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce

#### What information was incorrect, unhelpful, or incomplete?

In the Syntax section it describes the callbackFn as '_A function to execute on each element in the array (except for the first, if no initialValue is supplied)_.'. It's saying that if you don't provide an initialValue param and your array has items 0, 1, 2, and 3 then it will execute only three times - on indexes 1, 2, and 3. It will skip 0. It explicitly states this when describing the index:
_'The index of the current element being processed in the array. Starts from index 0 if an initialValue is provided. Otherwise, it starts from index 1'_.

#### Specific section or headline?

Syntax

#### What did you expect to see?

I was perfectly happy to accept this behaviour as described, but (see test section):

#### Did you test this? If so, how?

The page has it's own executor playground area and has code that contradicts the expected behaviour. I simply ran that code and it ran (WITHOUT initialValue and therefore should skip array element zero) contradictarily (new word, (c) me). It ran through the whole array, including element zero despite no initial value and accumulated array elements 0-3 giving a total value of 10. It would be 9 if the first element was skipped. Here's the code:

```
const array1 = [1, 2, 3, 4];
const reducer = (accumulator, currentValue) => accumulator + currentValue;

// 1 + 2 + 3 + 4
console.log(array1.reduce(reducer));     <<< no initial value supplied
// expected output: 10                           <<< but the expected output is somehow ten, why not 9?
```

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/array/reduce`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/array/reduce/index.html
* Last commit: https://github.com/mdn/content/commit/761ba0efe8230cec103437c215a79a2e6d2bb5f2
* Document last modified: 2021-06-15T08:30:35.000Z

</details>

### LINKED ISSUE #6497: Issue with "Array.prototype.reduce()": Confusing presentation of core concept
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce

#### What information was incorrect, unhelpful, or incomplete?
The fact that this doc (as pretty much any other info sources ou there, I am aware) calls the first argument of the reduce() method "accumulator" appeared to me as a strong suggestion that the returned value of the callback function applied to it was ADDED to the accumulator, as opposed to assigned to it. The fact that the very first demo of the method right on top of the page is that of the currentValue being added on to the acummulator, further compounds in my opinion the possibility for this specific confusion. I willingly admit that I also didn't help matters by not taking my time to read it carefully, but it just seems like that's a bad habit many people may unfortunately fall into when acessing this documentation.

#### Specific section or headline?
Mentioned above.

#### What did you expect to see?
Perhaps a different or aditional example in the demo on the top of the page to that which is already there. And also maybe, a strong highlighting of the word "assigned" in the sentence "Your reducer function's returned value is **_assigned_** to the accumulator, whose value is remembered across each iteration throughout the array, and ultimately becomes the final, single resulting value."

#### Did you test this? If so, how?

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/array/reduce`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/array/reduce/index.html
* Last commit: https://github.com/mdn/content/commit/761ba0efe8230cec103437c215a79a2e6d2bb5f2
* Document last modified: 2021-06-15T08:30:35.000Z

</details>

### PR DESCRIPTION
**Pull request by sideshowbarker**
We have multiple open issues — #6156 and #6497 — reporting that the current description we have in MDN for how `Array.prototype.reduce()` works is confusing, and asking that we refine it to make it more clear.

So this change replaces the current description by citing the https://tc39.es/ecma262/#sec-array.prototype.reduce description in the spec — which is much more clear — and then updating related text and variable names to match the spec language.

One key difference that helps make the spec language much more clear is that it doesn’t use the word “accumulator” but instead uses “previousValue”. So for consistency, this change also replaces “accumulator” throughout with “previousValue".

Fixes https://github.com/mdn/content/issues/6156
Fixes https://github.com/mdn/content/issues/6497

The change also drops the entire Polyfill section — in keeping with our current policy of not any longer maintaining polyfill code in MDN articles, but instead linking to polyfills at external sites (and in this case, the article already has a link to a high-quality polyfill).

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 1]
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
...
-1.  Accumulator
-2.  Current Value
-3.  Current Index
-4.  Source Array
+Perhaps the easiest-to-understand case for `reduce()` is to return the sum of all the elements in an array, by walking through the array element-by-element and accumulating a running sum of all the elements that precede each element, and then adding each element in turn to that accumulated sum — until there are no more elements to add — as shown in the following interactive example:
 
-Your **reducer** function's returned value is assigned to the accumulator,
-whose value is remembered across each iteration throughout the array, and ultimately
-becomes the final, single resulting value.
+{{EmbedInteractiveExample("pages/js/array-reduce.html")}}
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:22]: FYI This still uses the old accumulator terminology. Modify for consistency?

Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:22]: > FYI This still uses the old accumulator terminology. Modify for consistency?

Yes, thanks for reminding me — I plan on updating that interactive example after this PR is merged (but would rather not block this PR on that, because it’s a chicken/egg race condition; one or the other would need to land first, regardless).

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 2]
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
 ---
 {{JSRef}}
 
-The **`reduce()`** method executes a **reducer**
-function (that you provide) on each element of the array, resulting in a single output
-value.
+The **`reduce()`** method executes a “reducer” callback function (that you provide) on each element of the array, resulting in a single output value (the result of running the reducer to completion over the entire array).
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:18]:
```suggestion
The **`reduce()`** method executes a user-supplied “reducer” callback function on each element of the array, passing in the return value from the calculation on the preceding element. The final result of running the reducer across all elements of the array is a single value. 
```

I prefer the original text - the extra words don't help. What I'm trying to capture in this suggestion is that this was missing the concept that the results of the each call reflect the previous call. That's accumulation :-)

Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:18]: PS I don't love my words so not precious if you want to consider alternatives.

Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md]: > PS I don't love my words so not precious if you want to consider alternatives.

I quite like your words :) And I agree it’s an improvement — it reads a lot more smoothly, without the double-parentheticals.

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 3]
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
...
 
-{{EmbedInteractiveExample("pages/js/array-reduce.html")}}
-
-The **reducer** function takes four arguments:
-
-1.  Accumulator
-2.  Current Value
-3.  Current Index
-4.  Source Array
+Perhaps the easiest-to-understand case for `reduce()` is to return the sum of all the elements in an array, by walking through the array element-by-element and accumulating a running sum of all the elements that precede each element, and then adding each element in turn to that accumulated sum — until there are no more elements to add — as shown in the following interactive example:
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:20]: This sentence is too long/hard to parse. I'm having trouble rewriting it but minimally perhaps a full stop 
```suggestion
Perhaps the easiest-to-understand case for `reduce()` is to return the sum of all the elements in an array. 
The reducer walks through the array element-by-element, at each step adding the current array value to the result from the previous step (this result is the running sum of all the previous steps) — until there are no more elements to add.
This is shown in the following interactive example:
```

Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:20]: Yes. This is better for clarity and for breaking up the paragraph length for better readability.

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 4]
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
...
-
-- `initialValue` {{optional_inline}}
-  - : A value to use as the first argument to the first call of the
-    `callbackFn`. If no `initialValue` is
-    supplied, the first element in the array will be used as the initial
-    `accumulator` value and skipped as
-    `currentValue`. Calling `reduce()` on an empty array
-    without an `initialValue` will throw a {{jsxref("TypeError")}}.
+- *callbackFn*
+  - : A “reducer” function that takes four arguments:
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:49]: FWIW Don't know if we're supposed to use bullets or nested definition lists. I don't think it matters here.

Comment (hamishwillee) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:49]: @sideshowbarker But everywhere else parameters are in code style - e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 5]
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
...
+    - *previousValue* (the value resulting from the previous call to *callbackfn*)
+    - *currentValue* (the value of the current element)
+    - *currentIndex* {{optional_inline}}
+    - *array* (the array to traverse) {{optional_inline}}
 
-### Return value
+- *initialValue* {{optional_inline}}
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:55]: Ditto - should this be code style (and inline below)

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 6]
```diff
@@ -15,112 +15,83 @@ browser-compat: javascript.builtins.Array.reduce
...
+> It is a {{jsxref("TypeError")}} if the array contains no elements and *initialValue* is not provided.
+>
+> `reduce` does not directly mutate the object on which it is called but the object may be mutated by the calls to *callbackfn*.
+>
+> The range of elements processed by `reduce` is set before the first call to *callbackfn*. Elements that are appended to the array after the call to `reduce` begins will not be visited by *callbackfn*. If existing elements of the array are changed, their value as passed to *callbackfn* will be the value at the time `reduce` visits them; elements that are deleted after the call to `reduce` begins and before being visited are not visited.
 
 If the array only has one element (regardless of position) and no
-`initialValue` is provided, or if
-`initialValue` is provided but the array is empty, the solo value
+*initialValue* is provided, or if
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:87]: And I see you are being consistent within this doc.

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md (hunk 4 above)
Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:49]: > FWIW Don't know if we're supposed to use bullets or nested definition lists. I don't think it matters here.

Yeah it was definition list to start with, so just kept it that way.

Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:49]: > @sideshowbarker But everywhere else parameters are in code style - e.g. [developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)

Yeah, I don’t like that style, but I’ll change it here for consistency (and push the change in a minute)

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md (hunk 6 above)
Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:87]: > And I see you are being consistent within this doc.

😆

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md [hunk 7]
```diff
@@ -15,112 +15,83 @@ browser-compat: javascript.builtins.Array.reduce
...
 
-> **Note:** If `initialValue` is not provided,
-> `reduce()` will execute the callback function starting at index
-> `1`, skipping the first index. If `initialValue` is
-> provided, it will start at index `0`.
+## Description
 
-If the array is empty and no `initialValue` is provided,
-{{jsxref("TypeError")}} will be thrown.
+The ECMASCript spec describes the behavior of `reduce()` as follows:
```
Comment (hamishwillee) [inline files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:68]: ECMASCript  or ECMAScript ?

**Review by hamishwillee — APPROVED** [review]
So much better. Minor comments but is approved. Feel free to self merge if you want, or you can ping me and I will do it.

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md (hunk 7 above)
Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md]: > ECMASCript or ECMAScript ?

oofs, thanks — fixed

**Comment by sideshowbarker** [comment]
> So much better. Minor comments but is approved. Feel free to self merge if you want, or you can ping me and I will do it.

Please if you have time, give it a once-over to make sure I didn’t regress anything when I pushed the latest commit. And if looks right, please do go ahead and merge.

**Comment by hamishwillee** [comment]
@sideshowbarker Thank you for your 1E45 th contribution to MDN :-)

File: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md (hunk 1 above)
Comment (sideshowbarker) [reply files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:26]: > Yes, thanks for reminding me — I plan on updating that interactive example after this PR is merged

Opened https://github.com/mdn/interactive-examples/pull/1887
